netket/optimizer/qgt/qgt_jacobian_pytree.py

```diff
@@ -0,0 +1,229 @@
+########################################################################################
+
+
+@jax.jit
+def _matmul(
+    self: QGTJacobianPyTreeT, vec: Union[PyTree, Array]
+) -> Union[PyTree, Array]:
+    # Turn vector RHS into PyTree
+    if hasattr(vec, "ndim"):
+        _, unravel = nkjax.tree_ravel(self.params)
+        vec = unravel(vec)
+        ravel = True
+    else:
+        ravel = False
+
+    # Real-imaginary split RHS in R→R and R→C modes
+    reassemble = None
+    if self.mode != "holomorphic" and not self._in_solve:
+        vec, reassemble = nkjax.tree_to_real(vec)
+
+    if self.scale is not None:
+        vec = jax.tree_multimap(jnp.multiply, vec, self.scale)
+
+    result = mat_vec(vec, self.O, self.diag_shift)
+
+    if self.scale is not None:
+        result = jax.tree_multimap(jnp.multiply, result, self.scale)
+
+    # Reassemble real-imaginary split as needed
+    if reassemble is not None:
+        result = reassemble(result)
+
+    # Ravel PyTree back into vector as needed
+    if ravel:
+        result, _ = nkjax.tree_ravel(result)
+
+    return result
+
+
+@jax.jit
+def _split_matmul(self: QGTJacobianPyTreeT, vec: Array) -> Array:
+    pars = self.params
+    if self.mode != "holomorphic":
+        pars, _ = nkjax.tree_to_real(pars)
+
+    _, unravel = nkjax.tree_ravel(pars)
+    vec = unravel(vec)
+
+    if self.scale is not None:
+        vec = jax.tree_multimap(jnp.multiply, vec, self.scale)
+
+    result = mat_vec(vec, self.O, self.diag_shift)
+
+    if self.scale is not None:
+        result = jax.tree_multimap(jnp.multiply, result, self.scale)
+
+    result, _ = nkjax.tree_ravel(result)
+    return result
+
+
+@jax.jit
+def _solve(
+    self: QGTJacobianPyTreeT, solve_fun, y: PyTree, *, x0: Optional[PyTree] = None
+) -> PyTree:
+    # Real-imaginary split RHS in R→R and R→C modes
+    if self.mode != "holomorphic":
+        y, reassemble = nkjax.tree_to_real(y)
+
+    if self.scale is not None:
+        y = jax.tree_multimap(jnp.divide, y, self.scale)
+        if x0 is not None:
+            x0 = jax.tree_multimap(jnp.multiply, x0, self.scale)
+
+    # to pass the object LinearOperator itself down
+    # but avoid rescaling, we pass down an object with
+    # scale = None
+    # mode=holomoprhic to disable splitting the complex part
+    unscaled_self = self.replace(scale=None, _in_solve=True)
+
+    out, info = solve_fun(unscaled_self, y, x0=x0)
+
+    if self.scale is not None:
+        out = jax.tree_multimap(jnp.divide, out, self.scale)
+
+    # Reassemble real-imaginary split as needed
+    if self.mode != "holomorphic":
+        out = reassemble(out)
+
+    return out, info
+
+
+@jax.jit
+def _to_dense(self: QGTJacobianPyTreeT) -> jnp.ndarray:
+    pars = self.params
+
+    if self.mode != "holomorphic":
+        pars, _ = nkjax.tree_to_real(pars)
+
+    Npars = nkjax.tree_size(pars)
+    I = jax.numpy.eye(Npars)
+
+    return jax.vmap(self._split_matmul, in_axes=0)(I)
```

Comment:
This acts on the **rows** of the identity matrix rather than its columns, so you get the transpose of the S matrix, which is different from the S matrix in the `holomorphic` case. You would need `in_axes=1, out_axes=1`. 
But in any case, `PyTree::todense` should be replaced with https://github.com/netket/netket/pull/674/files/6b7145a4cfabb209d77559a4c009452dcef50c9b#r632923168 to improve performace